File: Python/第四章/03-plan-replan-agent/deepseek_client.py
"""DeepSeek Chat Completions 客户端。"""

import json
import os
import time
import urllib.error
import urllib.request

API_URL = os.getenv("DEEPSEEK_BASE_URL") or "https://api.deepseek.com/chat/completions"
MODEL = os.getenv("DEEPSEEK_MODEL") or "deepseek-v4-flash"


async def call_deepseek(*, messages, json_output=False, max_tokens=2400, env=None, transport=None):
    """调用 DeepSeek Chat Completions。

    Planner 和 Replanner 使用 JSON Output，最终结论使用普通文本输出。
    """

    env = env or os.environ
    api_url = env.get("DEEPSEEK_BASE_URL") or API_URL
    model = env.get("DEEPSEEK_MODEL") or MODEL
    api_key = env.get("DEEPSEEK_API_KEY")

    if not api_key:
        raise RuntimeError("缺少 DEEPSEEK_API_KEY，请先在 .env 中完成配置。")

    started_at = time.perf_counter()
    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json",
    }
    payload = {
        "model": model,
        "messages": messages,
        "thinking": {
            "type": "disabled",
        },
        "temperature": 0.1,
        "max_tokens": max_tokens,
    }

    if json_output:
        payload["response_format"] = {
            "type": "json_object",
        }

    if transport is None:
        status, data = _post_json(api_url, headers, payload)
    else:
        status, data = transport(api_url, headers, payload)

    if status < 200 or status >= 300:
        raise RuntimeError(f"DeepSeek 调用失败：{status} {json.dumps(data, ensure_ascii=False)}")

    choices = data.get("choices") or []
    message = choices[0].get("message") if choices else None

    if not message:
        raise RuntimeError(f"DeepSeek 没有返回有效消息：{json.dumps(data, ensure_ascii=False)}")

    return {
        "message": message,
        "latencyMs": round((time.perf_counter() - started_at) * 1000),
        "usage": data.get("usage"),
    }


def _post_json(api_url, headers, payload):
    request = urllib.request.Request(
        api_url,
        data=json.dumps(payload, ensure_ascii=False).encode("utf-8"),
        headers=headers,
        method="POST",
    )

    try:
        with urllib.request.urlopen(request, timeout=60) as response:
            return response.status, json.loads(response.read().decode("utf-8"))
    except urllib.error.HTTPError as error:
        body = error.read().decode("utf-8")

        try:
            data = json.loads(body)
        except json.JSONDecodeError:
            data = {"message": body}

        return error.code, data

File: Node/第四章/03-plan-replan-agent/deepseek-client.js
const API_URL =
	process.env.DEEPSEEK_BASE_URL || 'https://api.deepseek.com/chat/completions'

const MODEL = process.env.DEEPSEEK_MODEL || 'deepseek-v4-flash'

/**
 * 调用 DeepSeek Chat Completions。
 *
 * Planner 和 Replanner 使用 JSON Output，最终结论使用普通文本输出。
 *
 * @param {object} input 调用参数
 * @param {Array} input.messages 当前消息
 * @param {boolean} [input.jsonOutput=false] 是否要求返回 JSON
 * @param {number} [input.maxTokens=2400] 最大输出 Token
 * @returns {Promise<object>} 模型消息和调用统计
 */
async function callDeepSeek({
	messages,
	jsonOutput = false,
	maxTokens = 2400
}) {
	if (!process.env.DEEPSEEK_API_KEY) {
		throw new Error('缺少 DEEPSEEK_API_KEY，请先在 .env 中完成配置。')
	}

	const startedAt = Date.now()
	const response = await fetch(API_URL, {
		method: 'POST',
		headers: {
			Authorization: `Bearer ${process.env.DEEPSEEK_API_KEY}`,
			'Content-Type': 'application/json'
		},
		body: JSON.stringify({
			model: MODEL,
			messages,
			...(jsonOutput
				? {
						response_format: {
							type: 'json_object'
						}
					}
				: {}),
			thinking: {
				type: 'disabled'
			},
			temperature: 0.1,
			max_tokens: maxTokens
		})
	})

	const data = await response.json()

	if (!response.ok) {
		throw new Error(
			`DeepSeek 调用失败：${response.status} ${JSON.stringify(data)}`
		)
	}

	const message = data.choices?.[0]?.message

	if (!message) {
		throw new Error(`DeepSeek 没有返回有效消息：${JSON.stringify(data)}`)
	}

	return {
		message,
		latencyMs: Date.now() - startedAt,
		usage: data.usage
	}
}

module.exports = {
	MODEL,
	callDeepSeek
}

